src/components/esn-app-grid/esn-app-grid.tsx

```diff
@@ -0,0 +1,80 @@
+import { Component, Prop, ComponentInterface, State, Host, h } from '@stencil/core';
+import fetchSvgIcon from '../../utils/fetch-icon';
+import { InternalItem } from './esn-app-grid.types';
+import { AppGridTogglerIcon } from './app-grid-icon/AppGridIcon';
+
+@Component({
+  tag: 'esn-app-grid',
+  styleUrl: 'esn-app-grid.scss',
+  shadow: true
+})
+export class EsnAppGrid implements ComponentInterface {
+  private popover: HTMLElement;
+
+  @State() internalItems: InternalItem[] = [];
+
+  /**
+   * A serialized JSON object that contains information about the available apps.
+   */
+  @Prop() serializedItems: string;
+
+  componentWillLoad() {
+    try {
+      if (!this.serializedItems) {
+        throw new Error("The 'serializedItems' attribute/property must be set to a serialized JSON");
+      }
+
+      const items = JSON.parse(this.serializedItems);
+
+      if (Array.isArray(items) && items.length) {
+        this.internalItems = items.map(({ name, url }) => ({
+          name,
+          url
+        }));
+      } else {
+        throw new Error("The parsed 'serializedItems' has an invalid JSON structure");
+      }
+
+      Promise.all(items.map(({ iconUrl }) => fetchSvgIcon(iconUrl))).then(iconContents => {
```

Comment:
Now that we are doing cool ES code, can't we async/await?

Reply:
Yes but we'll need to wrap it a smaller function inside the `componentWillLoad` method since if you use `async componentWillLoad`, it will wait until the asynchronous action is finished before rendering the component, which is actually not what we want. If we do that, it might not be cleaner than using `promise.then`.

Reply:
Speaking of which, I've forgotten to include a fallback loading spinner while the icons are loading. Added to TODOs.